core/src/main/java/com/linecorp/armeria/internal/common/LeakTracingRequestContextStorage.java

```diff
@@ -0,0 +1,133 @@
+import static java.util.Objects.requireNonNull;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.ClientRequestContextWrapper;
+import com.linecorp.armeria.common.RequestContext;
+import com.linecorp.armeria.common.RequestContextStorage;
+import com.linecorp.armeria.common.RequestContextWrapper;
+import com.linecorp.armeria.common.annotation.Nullable;
+import com.linecorp.armeria.common.annotation.UnstableApi;
+import com.linecorp.armeria.common.util.Sampler;
+import com.linecorp.armeria.server.ServiceRequestContext;
+import com.linecorp.armeria.server.ServiceRequestContextWrapper;
+
+/**
+ * A {@link RequestContextStorage} which keeps track of {@link RequestContext}s, reporting pushed thread
+ * information if a {@link RequestContext} is leaked.
+ */
+@UnstableApi
+final class LeakTracingRequestContextStorage implements RequestContextStorage {
+
+    private final RequestContextStorage delegate;
+    private final Sampler<? super RequestContext> sampler;
+
+    /**
+     * Creates a new instance.
+     * @param delegate the underlying {@link RequestContextStorage} that stores {@link RequestContext}
+     * @param sampler the {@link Sampler} that determines whether to retain the stacktrace of the context leaks
+     */
+    LeakTracingRequestContextStorage(RequestContextStorage delegate,
+                                     Sampler<? super RequestContext> sampler) {
+        this.delegate = requireNonNull(delegate, "delegate");
+        this.sampler = requireNonNull(sampler, "sampler");
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T push(RequestContext toPush) {
+        requireNonNull(toPush, "toPush");
+        if (sampler.isSampled(toPush)) {
+            return delegate.push(warpRequestContext(toPush));
+        }
+        return delegate.push(toPush);
+    }
+
+    @Override
+    public void pop(RequestContext current, @Nullable RequestContext toRestore) {
+        requireNonNull(current, "current");
+        delegate.pop(current, toRestore);
+    }
+
+    @Nullable
+    @Override
+    public <T extends RequestContext> T currentOrNull() {
+        return delegate.currentOrNull();
+    }
+
+    @Override
+    public RequestContextStorage unwrap() {
+        return delegate;
+    }
+
+    private static RequestContextWrapper<?> warpRequestContext(RequestContext ctx) {
+        return ctx instanceof ClientRequestContext ?
+               new TraceableClientRequestContext((ClientRequestContext) ctx)
+               : new TraceableServiceRequestContext((ServiceRequestContext) ctx);
+    }
+
+    private static String stacktraceToString(StackTraceElement[] stackTrace, String threadName,
+                                             RequestContext unwrap) {
+        final StringBuilder builder = new StringBuilder(512);
+        builder.append(unwrap).append(System.lineSeparator())
+               .append("At thread [").append(threadName)
```

Comment:
I don't think we need `threadName` because it's the same thread that raises `newIllegalContextPoppingException `. (If the thread is changed, the exception is not raised. 😉 )

Reply:
Ok, I got it. Thanks